src/Refactoring-Core/RBInlineTemporaryRefactoring.class.st

```diff
@@ -38,55 +38,47 @@ RBInlineTemporaryRefactoring class >> model: aRBSmalltalk inline: anInterval fro
 ]
 
 { #category : 'preconditions' }
-RBInlineTemporaryRefactoring >> applicabilityPreconditions [
-
+RBInlineTemporaryRefactoring >> applicabilityPreconditions [ 
+	
 	^ {
-		  (RBCondition definesSelector: selector in: class).
-		  (RBCondition withBlock: [
-			   self checkSelectedInterval.
-			   true ]) }
-]
-
-{ #category : 'preconditions' }
-RBInlineTemporaryRefactoring >> checkSelectedInterval [
-
-	sourceTree := class parseTreeForSelector: selector.
-	sourceTree ifNil: [ self refactoringError: 'Could not parse source' ].
-	assignmentNode := sourceTree whichNodeIsContainedBy: sourceInterval.
-	assignmentNode isAssignment
-		ifFalse: [ self refactoringError: 'The selected node is not an assignment statement' ].
-	definingNode := assignmentNode whoDefines:
-		                assignmentNode variable name.
-	self hasOnlyOneAssignment
-		ifFalse: [ self refactoringError: 'There are multiple assignments to the variable' ].
-	( RBReadBeforeWrittenTester
-		isVariable: assignmentNode variable name
-		writtenBeforeReadIn: definingNode )
-		ifFalse: [ self refactoringError: 'The variable is possible read before it is assigned' ]
+		(ReDefinesSelectorsCondition new definesSelectors: (Array with: selector) in: class).
+		(ReIsValidInstanceVariableName new name: assignmentNode variable name; check).
+		(ReBlockCondition new 
+			block: [
+				(ReSingleAssignmentCondition new 
+					initialize: assignmentNode name parseTree: self parseTreeSearcher; check).
+				(ReVariablesNotReadBeforeWrittenCondition new 
+					variables: {assignmentNode variable name}; subtree: sourceTree; check).
+				true]).
+	}
 ]
 
 { #category : 'transforming' }
 RBInlineTemporaryRefactoring >> compileMethod [
 	class compileTree: sourceTree
 ]
 
-{ #category : 'testing' }
-RBInlineTemporaryRefactoring >> hasOnlyOneAssignment [
-	| searcher |
-	searcher := self parseTreeSearcher.
-	searcher
-		matches: assignmentNode variable name , ' := ``@object'
-		do: [ :aNode :answer | answer + 1 ].
-	^ (searcher executeTree: definingNode initialAnswer: 0) == 1
-]
-
 { #category : 'initialization' }
 RBInlineTemporaryRefactoring >> inline: anInterval from: aSelector in: aClass [
 	class := self classObjectFor: aClass.
 	selector := aSelector.
 	sourceInterval := anInterval
 ]
 
+{ #category : 'transforming' }
+RBInlineTemporaryRefactoring >> prepareForExecution [
+	
+	sourceTree := class parseTreeForSelector: selector.
+	sourceTree ifNil: [ self refactoringError: 'Could not parse source' ].
+	
+	assignmentNode := sourceTree whichNodeIsContainedBy: sourceInterval.
+	assignmentNode isAssignment
+		ifFalse: [ self refactoringError: 'The selected node is not an assignment statement' ].
```

Comment:
I'm wondering if this should be a precondition or not. We can leave it like this, but it is something to think about and return to in the future.